### PR TITLE
feat(core): suspension store, records, and signed resume tokens

### LINKED ISSUE #549: feat(core): suspension store, records, and signed resume tokens
## Summary

Slice 1 of #417. The persistence and identity layer for parked exchanges. No DSL, no executor changes, no user-visible surface: this is the foundation the other slices build on.

Design lives in #417. This issue is the buildable slice.

## Scope

**The record.** `Suspension` as specified in #417, in `@routecraft/routecraft`. Not in `@routecraft/ai`: the store is shared by plain capabilities, approvals (#416), and the agent tier (#258), and building an agent-shaped one first is the mistake this whole realignment corrects.

```ts
interface Suspension {
  id: string
  routeId: string
  position: number            // resume at N+1
  continuationHash: string    // steps N+1..end plus the expect schema
  exchange: SerializedExchange
  expect: StandardSchemaV1    // or a reference to it
  stepState?: unknown         // opaque, owned by the suspending step
  actionFingerprint: string
  suspendedAt: Date
  expiresAt?: Date
  status: 'suspended' | 'resumed' | 'expired' | 'denied'
  terminal?: SerializedOutcome // cached, for idempotent re-resume
  resumedBy?: PrincipalRef
  resumedAt?: Date
}
```

**The store contract**, with two backends: in-memory for tests and ephemeral use, sqlite as the durable default where the runtime allows (see the runtime question below). Postgres and redis are explicitly out of scope; the contract is shaped so they can be added later without touching call sites.

**Signed resume tokens.** Mint and verify. Single use, bound to a suspension id. Minting must be possible *before* the suspend step executes, since `ex.suspension.resumeUrl` has to be readable by the notification step that precedes it (the key is the exchange id).

**`continuationHash` computation.** Over steps N+1 to the end plus the `expect` schema. Not the whole pipeline. See #417's "Pipeline compatibility" for why, including the recorded limit: the hash covers the tail's step definitions only, never the behaviour of what the tail calls.

**`actionFingerprint` computation.** Hash over routeId, position, `continuationHash`, serialized body. Audit-grade binding of an approval to the operation it authorized.

## Non-obvious requirements

**`markResumed` must be an atomic compare-and-swap that reports whether it won.** Multi-node coordination is out of scope for the epic, but a store whose resume is a read-then-write cannot be made safe later without a rewrite. Costs nothing now.

**The sqlite default, narrowed (decide here, first task).** `bun:sqlite` alone cannot back the default: core targets Node 22+ as well as Bun.

The version matrix matters more than the API's existence, and it rules out the obvious answer. `node:sqlite` landed in 22.5.0 **behind `--experimental-sqlite`**, and only runs unflagged in later majors. Core's engines floor is Node 22.0+, so on the supported Node range the native module either does not exist (below 22.5) or needs a CLI flag the framework cannot set for the user. That inverts the naive preference order for the Node arm.

Candidates, evaluated against that matrix:

1. Split by runtime: `bun:sqlite` under Bun (always available on our floor); under Node, `better-sqlite3` as an optional peer through `loadOptionalPeer` with RC5017 and an install hint, per the house rule in `.standards/ci-cd.md` § 6, falling back to in-memory with a loud warning when the peer is absent.
2. In-memory default in core everywhere, sqlite (per arm as above) as the first opt-in backend.
3. `node:sqlite` for the Node arm: **deferred**, graduating to preferred only once the engines floor moves past the version where it runs unflagged. Not viable on a 22.0+ floor.

Whichever wins, the decision, the version matrix it was evaluated against, and the graduation condition for `node:sqlite` go in the store's JSDoc.

**Schema migration is owned by the store implementation.** No foot-guns on first run.

## Acceptance criteria

- [ ] `Suspension` and the store contract exported from `@routecraft/routecraft`.
- [ ] In-memory backend passes a shared contract test suite.
- [ ] Sqlite backend passes the same contract test suite on both Bun and Node 22 (via the chosen per-runtime strategy), and the absent-peer fallback path is tested.
- [ ] `markResumed` is atomic and returns whether the caller won the race; a concurrent double-resume test proves only one wins.
- [ ] Token mint and verify, single use, and mintable before the suspend step runs.
- [ ] `continuationHash` covers steps N+1 onward only; a test proves an edit before the suspend point does not invalidate a parked exchange.
- [ ] `actionFingerprint` recorded on every suspension.
- [ ] The sqlite-runtime decision recorded in the store's JSDoc, including the version matrix and the `node:sqlite` graduation condition.
- [ ] Schema migration handled by the store on first run.
- [ ] Tests per `.standards/testing.md`; `bun run all` green.

## Out of scope

* `.suspend()` / `.resume()` operations and executor integration. #550.
* TTL, expiry sweeping, startup scan. #551.
* Postgres and redis backends. Once the contract is stable.
* What goes in `stepState`. Owned by each suspending step; the agent tier's version is #268.

## Related

* Epic: #417
* #526 adapter sandbox and `Secret<T>`: secrets must never reach the store
* #355 principal authenticity: a rehydrated principal must be re-marked as restored

### PR DESCRIPTION
**Pull request by ex0b1t**
Closes #549. Slice 1 of 3 in the suspend/resume lane (#417); #550 and #551 follow on their own branches.

The persistence and identity layer for parked exchanges. No DSL and no executor changes: `.suspend()` / `.resume()` are #550, TTL and the sweeper are #551. The store lives in core rather than `@routecraft/ai` because it is shared by plain capabilities, human approvals (#416) and the agent tier (#258) alike; `stepState` is the one opaque slot the store never interprets, which is what lets the agent tier adopt this store instead of growing a second one.

`bun run all` green (2314 tests). Cross-runtime suite green under both Bun and Node.

## What landed

**The contract.** `SuspensionStore` is async throughout and its state transitions are compare-and-swap rather than read-then-write, so a resume racing the sweeper has exactly one winner and multi-node coordination stays addable later without a rewrite. Postgres and redis are out of scope but the contract is shaped so they drop in without touching call sites.

**Two backends, one shared contract suite.** The suite runs three times (memory, sqlite in-process, sqlite on disk) and every backend satisfies it identically. The sqlite backend runs a per-runtime driver split: `bun:sqlite` under Bun, `better-sqlite3` as a new optional peer under Node via `loadOptionalPeer` with `RC5017`. `node:sqlite` is deferred; the version matrix behind that call and its graduation condition (engines floor past Node 24) are recorded in the driver JSDoc, per the ticket's acceptance criteria. A cross-runtime suite proves the split on both runtimes.

**Signed resume tokens.** HMAC-SHA256, base64url, mintable from a suspension id alone so a notification step *ahead of* the suspend can send a working link. The signing secret is required config from `ROUTECRAFT_SUSPENSION_SECRET` or `suspension: { secret }`, never generated into the store, and its absence is a build-time `RC5040` rather than a surprise on the first payout.

**Store location and secret are both env-configurable**, as the showcase lane needs: it mounts the store on a Docker volume and injects the secret via compose.

## Security findings found by review, verified by execution, and fixed

The six-lens internal review ran before this PR. Three findings were reproduced by running the code, not just reading it, and all three would have been silent in production:

1. **`__proto__` prototype hijack in the serializer.** Any body from `JSON.parse` can carry a literal `__proto__` key. Assigning it onto an object literal hits the inherited setter: the field silently vanished between suspend and resume, and the serialized object carried an attacker-supplied prototype for the rest of the suspend path. Fixed with a null-prototype accumulator on encode and `defineProperty` on decode. This file is a declared security boundary, so an accumulator honouring that setter was the wrong primitive regardless of whether slice 1 has a consumer yet.
2. **Forgeable `Date` envelope.** A body field shaped like the internal envelope revived as a `Date` after resume. On a suspending route the body is routinely attacker-influenced (a webhook payload, an inbound mail), so this was type confusion at exactly the point the code trusts the round trip, and `actionFingerprint` could not detect it because it hashes the encoded form. The envelope key is now reserved at suspend time, and a corrupt stored envelope fails loudly instead of reviving as `Invalid Date`.
3. **Symbol-keyed properties dropped, not refused.** Invisible to `Object.entries`, so they parked successfully and the data was simply gone after resume, which is the failure mode the whole encode walk exists to prevent. Symbol keys are the framework's own idiom for attaching state, so an adapter author moving a value onto the exchange hit this with no signal.

## Hash scope fix

`continuationHash` originally hashed only the adapter's **callables**. A class-based adapter keeps its config in a data property and exposes a callable whose source text is identical for every instance, so `http({ url: bankA })` and `http({ url: bankB })` hashed identically: a parked approval could resume into a payment to a different payee and pass the compatibility check. That undercuts the epic's headline property, so the digest now covers adapter options too, with unhashable state (live clients, sockets) collapsing to a placeholder so such a step stays comparable. Tests cover both directions, including that widening the hash did not cost the "edit before the suspend point stays resumable" property it exists for.

## Other review fixes

- The compare-and-swap trusted the driver's `run()` return value. `bun:sqlite` only began reporting `changes` partway through the 1.1 line and our declared floor is 1.1.0, so on an in-range Bun every transition would have reported *lost* while actually succeeding, and nothing would ever have resumed. Now read via `SELECT changes()`.
- `busy_timeout` left at the driver default, which is 0 under Bun and 5s under Node: a second writer on the store file failed a resume instantly on one runtime and waited on the other, so the bug would not reproduce for anyone debugging on the other.
- The newer-schema downgrade guard sat after the up-to-date check and was unreachable.
- `BEGIN` outside the try and an unguarded `ROLLBACK` masked both busy locks and corrupt-column parse failures.
- Store failures collapsed into `RC5001`, which is **retryable**: a duplicate id could burn a retry budget on an insert that can never succeed. Added `RC5044`, not retryable, discriminating the collision.
- Token verification hand-rolled the constant-time compare instead of using the existing `timingSafeStringEqual`, leaving the resume-token path outside the helper built so that comparison exists exactly once.
- Any-strength secrets were accepted; a token holder can guess offline without limit, so there is now a 32-byte floor with a generation command in the error.
- `allowEphemeralSecret` sat on the public config type and short-circuited the named-environment gate `.standards/security.md` §6a requires. Moved to a separate internal seam type `defineConfig` cannot reach.
- Teardown closed user-supplied stores it did not own, and reported them as the `sqlite` backend on the one field that answers whether a deployment is durable.
- `MemorySuspensionStore.close()` wiped its records while sqlite kept them, so the two were not substitutable.
- Added `purgeSettled` to the contract: nothing reclaimed settled records, and the memory backend is also the Node fallback, so a long-running process accumulated every exchange that ever suspended.

## Notes for downstream lanes

- **The engine surface is not published.** Hashing, serialization, token minting and runtime resolution stay behind `src/suspension/index.ts`, which is where #550's executor imports them from. Only config, the store contract, the two backends and the env-var names reach the package index, per `.standards/api-stability.md`.
- **Event names.** The epic writes `exchange:suspended`; the fixed registry's convention is `route:exchange:*`, so #550 will emit `route:exchange:suspended` / `:resumed` / `:expired`. The epic's actual constraint, no parallel `agent:*` set, is unaffected.
- **The `Secret` rule is currently inert.** #526 has not landed, so there is no `Secret` to refuse. `BRAND.Secret` is reserved and the serializer already refuses anything carrying it, so the rule goes live the moment #526 brands its wrapper. Tested against a hand-branded object.
- **`RC5043` is a new `authorize()` rejection.** A principal rehydrated from a suspension is marked *restored*, never re-marked authentic (#355's bug class, and suspension is the most direct way to reintroduce it). It is distinct from `RC5023` because the caller's fix differs: re-verify, do not mint.
- **Build-sensitive hashing, accepted for v1.** `continuationHash` reads step identity from function source, a property of the build rather than the source tree, so a deploy changing emitted text can invalidate parked exchanges whose code did not change. The failure direction is safe (re-ask through the error channel, never a wrong resume), running TypeScript directly is unaffected, and the caveat is documented for users on the configuration reference. An author-declared version escape hatch is recorded as a possible future enhancement on #417.

## Deferred

Two reuse findings, filed rather than folded in because both touch telemetry and the CLI: #561 (one sqlite driver resolver, plus widening the DoD optional-peer rule beyond adapters; carries a telemetry behaviour decision) and #562 (extract the thrice-declared sqlite typing).

One reported finding was **dropped**: `better-sqlite3` is not bundled into `@routecraft/cli`, because `packages/cli/src/runtime-gate.ts` hard-refuses to start under Node, so `bun:sqlite` is always available there and the failure scenario cannot occur.

---
_Generated by [Claude Code](https://claude.ai/code/session_011AEv7bppFMvw499Pak6FGT)_

---
## Summary by cubic
Adds the suspension persistence layer and signed resume tokens for parked exchanges, with a strict hashing/serialization boundary and memory/SQLite backends. This is slice 1 of #417 and lays the groundwork for `.suspend()`/`.resume()` without DSL or executor changes.

- **New Features**
  - Async `SuspensionStore` with compare-and-swap; backends: in-memory and SQLite (driver split `bun:sqlite` on Bun, optional peer `better-sqlite3` on Node); includes `purgeSettled`. Config via `suspension.store` or `ROUTECRAFT_SUSPENSION_STORE`.
  - Signed resume tokens (HMAC‑SHA256, base64url) with a 32‑byte min secret from `ROUTECRAFT_SUSPENSION_SECRET` or `suspension.secret`; misconfig is a startup error (RC5040–RC5041). Uses the shared timing‑safe compare.
  - Continuation hashing for tail steps plus the expect schema; now covers factory args and nested callback source so adapters can’t swap destinations by accident.

- **Bug Fixes**
  - Serialization boundary: only plain JSON is persisted; fixes `__proto__` hijack; reserves the Date envelope; normalizes `undefined`/holes/`-0`; refuses Secret‑branded values; rejects Dates with own props or subclasses; rejects arrays with named/symbol/hidden props; now also refuses hidden own properties on plain objects to prevent silent loss; principals restored from storage are marked and rejected by `authorize()` (RC5043). Restores a linear‑cost hidden‑props check.
  - SQLite store: do not mask persistence-boundary errors (propagate RC5042 instead of misreporting RC5044); use `SELECT changes()`; consistent `busy_timeout`; transactional, race‑free migrations; close leaked connections on failures; define negative‑limit handling; classify driver errors (duplicate id RC5044 non‑retryable; `SQLITE_BUSY` RC5045 retryable); accurate backend reporting and teardown.
  - Tokens/store: token id always appends the sequence with `~`; secrets below 32 bytes refused; blank `ROUTECRAFT_SUSPENSION_STORE` treated as unset; memory store clones records and builds rows field‑by‑field from `NewSuspension` so pre‑settled fields cannot leak.
  - Continuation hashing: hash step source verbatim (no whitespace normalization) to avoid ASI/regex/comment collisions; behavior is now formatting and line‑ending sensitive, and docs call this out with guidance to pin line endings.

<sup>Written for commit 8779d4f9684ca0ed0ce942025569efbd6c148116. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/routecraftjs/routecraft/pull/563?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by ex0b1t** [comment]
Second review pass applied in 1570ed5. `bun run all` green, cross-runtime green under Bun and Node.

The most important one invalidates a claim this PR made earlier, so calling it out plainly.

**`continuationHash` did not cover real adapters.** The previous commit widened the digest to include an adapter's own data properties and I proved it with a hand-built class. That was the wrong fixture. Every factory-built adapter returns a role facade (`{ adapterId, subscribe, send, fetch }`) whose own properties are bound methods with byte-identical source for every instance, so the walk reached nothing: `file({ path: a })` and `file({ path: b })` still produced the same hash, and so did two different dynamic path callbacks. A parked approval could resume into a write to a different destination, which is exactly the property the hash exists to protect.

Fixed by folding in the factory arguments `tagAdapter` already records, and by hashing callback source nested inside options rather than collapsing it to a placeholder. The new tests drive a real `file()` adapter, which is what would have caught it the first time.

Also fixed:

| Area | Change |
|---|---|
| hash | Whitespace normalization no longer collapses runs inside string literals, so `pay("acct  123")` and `pay("acct 123")` stop hashing alike |
| store | `user_version` is read inside the migration transaction; two processes starting on one file no longer race into a failed migration and a memory fallback |
| store | A failed PRAGMA or migration closes the connection instead of leaking a locked handle |
| store | `SQLITE_BUSY` after the busy timeout is now `RC5045`, retryable, distinct from the permanent `RC5044` |
| store | `findExpired`'s `limit` contract is defined and enforced by both backends, instead of meaning "unbounded" in sqlite and "drop from the end" in memory |
| serialize | `stepState` and terminal bodies go through the same JSON-data encoder, so sqlite and memory cannot disagree about them |
| serialize | `undefined` and holes in arrays normalize to `null` at encode time, matching what a JSON round trip does |
| serialize | A `Date` with own properties, and an array with named or symbol-keyed properties, are refused rather than resuming with that state gone |
| config | `suspensionPlugin` no longer accepts the test seams on its public signature |

Two tests were wrong and are corrected: the undefined-semantics test asserted the opposite of its own documented expectation, and the whitespace test compared token-identical functions, so it never exercised the property it named. Both were fair catches.

**Not fixed, and why.** The note about `ContextBuilder.build()` rethrowing without unwinding already-applied plugins is real, but it is not specific to this slice: any plugin holding a resource has the same exposure, and the fix belongs in the builder's lifecycle rather than in one plugin's teardown. Adding a bespoke unwind here would hide a framework-wide gap behind a local workaround. Worth its own issue against the builder if it is not already tracked.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ex0b1t** [comment]
Third review pass applied in 390c073. `bun run all` green, cross-runtime green under Bun and Node.

Eleven of the twelve findings taken. Three were more than cosmetic:

| Finding | Why it mattered |
|---|---|
| Blank `ROUTECRAFT_SUSPENSION_STORE` counted as explicit config | A context with the variable set to `""` tried to open the working directory as a database and failed to start. Blank now means unset, matching how a blank signing secret is already treated |
| `#transition` stored without cloning | The memory backend held caller-owned values (the resume timestamp, the principal ref) by reference, so the clone-in/clone-out invariant the class documents did not actually hold |
| `SuspensionTestSeams` on the package entry point | The Definition of Done forbids `@internal` symbols there, and the type's own doc says users must not have it. It stays behind the intra-package barrel now; `@routecraft/testing` declares the shape it supplies |

Also fixed: only `create` and `#transition` classified driver errors, so read paths surfaced raw driver errors and a corrupt column surfaced a raw `SyntaxError`; the unparseable-date-envelope error read "Cannot suspend: exchange stored exchange holds ..." on what is a resume path; and the 32-byte floor comment claimed to match the SHA-256 *block* size, which is 64 bytes, when 32 is the digest size and the actual RFC 2104 basis for the floor.

Test fixes: the cross-runtime suite closes stores from a lifecycle hook instead of the last line of each test, so a failed assertion cannot leave a handle open and turn teardown into the reported error; the shared contract suite now covers the `findExpired` limit rejection both backends enforce, which was previously guarded but untested; the runtime tests isolate themselves from an inherited store environment variable; and the empty-continuation test asserts a digest shape rather than comparing a call to itself.

**Not taken:** widening the `better-sqlite3` peer range to majors 12 and 13. The reasoning is sound on API surface, but we test against 11 only, and declaring support for majors nothing exercises is a promise the suite does not back. Worth doing deliberately, with the cross-runtime job running against the widened range, rather than as a range bump here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ex0b1t** [comment]
Second review round triaged against the code rather than taken at face value. Nine of ten taken, one declined. Every accepted finding was reproduced with a probe before the fix and has a regression test that fails on the previous behaviour. Pushed in `702723a`.

### Taken

**`hash.ts` `describable` drops a `__proto__` own key (P1).** Real, and it is the same bug class already fixed in `serialize.ts` and missed here. `Object.entries` sees `__proto__` when it came from `JSON.parse` (a config file, an upstream payload), and assigning it onto an object literal hits the `Object.prototype` setter, so the field vanished from the projection. A tail option under that key could change under a parked approval without moving the digest. Fixed with the same null-prototype accumulator.

**`hash.ts` ASI hazard (P1).** Real. `return` + newline + `value` returns undefined; `return value` does not. Collapsing all whitespace made the two hash identically. A run containing a line terminator now collapses to a single `\n` instead of a space. CRLF and LF still normalise to the same thing, so the line-ending case the collapsing existed for still holds; a reformat that moves line boundaries now moves the hash, which is the over-invalidating direction the file's own contract prefers. The whitespace test was rewritten to assert what actually holds, and the ASI pair is now a test of its own.

**`tokens.ts` `suspensionIdFor` collision (P1).** Real. `("a#1", 0)` and `("a", 1)` both produced `"a#1"`, and two unrelated parked exchanges sharing an id means one overwrites the other in the store. The sequence is now always appended. Separately, `#` was a poor separator regardless: it delimits a URL fragment, so an id would have been truncated on its first trip through a browser. Moved to `~`, unreserved in RFC 3986.

**`serialize.ts` Date subclass (P2).** Real: accepted, then resumed as a plain `Date` with its added behaviour gone. It now falls through to the class-instance refusal.

**`serialize.ts` non-enumerable own property on a Date (P2).** Real. `Object.keys` misses it, so it was accepted and lost. Reads `getOwnPropertyNames` now.

**`serialize.ts` digit-only non-index array keys (P2).** Real. `"01"` and any out-of-range numeric pass `/^\d+$/`, are never visited by `Array.from`, and were dropped in silence. Array keys are now tested for being positions the index walk actually reaches.

**`types.ts` create-input type (P2).** Taken, slightly wider than suggested: `create()` takes a `NewSuspension` that omits `status` and every settlement field, not just pins `status` to `"suspended"`. A record inserted already settled is one that every compare-and-swap then refuses to move, so the exchange parks forever with nothing raised.

**Concurrency test does not create a real race (P2).** This is the most useful finding in the round and it is correct. Both drivers are synchronous, `#transition` runs BEGIN/UPDATE/COMMIT in one event-loop turn, and the two `markResumed` calls under `Promise.all` could not interleave. The assertion would have passed against an implementation with no atomicity at all, which matters because #549's acceptance criteria rest on this proof.

Replaced with a genuine race: four OS processes, each opening its own connection to one on-disk database, released together by a wall-clock barrier. Verified by temporarily swapping `markResumed` for a read-then-write with a 50ms gap, against which the new test reports **four winners**; the real implementation reports one. The cross-runtime test keeps its per-driver check but is reworded to claim only what it proves (the transition is one-shot) and points at where the race lives, since vitest under both runtimes is the wrong place to spawn children.

**Runtime test environment isolation (P2).** Real flake. An ambient `ROUTECRAFT_SUSPENSION_SECRET` turns the ephemeral-key test into an env-key test. Both suspension environment variables are now saved and cleared in `beforeEach` for the whole suite.

### Declined

**`SUSPENSION_RUNTIME` on the entry point exposes the signer (P3).** Not a boundary. Anything holding a `CraftContext` is in-process application code that can already read `ROUTECRAFT_SUSPENSION_SECRET` from `process.env` or the value it passed to `defineConfig`; hiding the store key buys nothing against that reach. The key is public because operator tooling legitimately needs the store (a custom sweeper, a pending-suspension report), and #551 builds on that. The engine-only machinery does stay behind `./suspension/index.ts`.

`bun run all` green (2333 tests), cross-runtime green under both Bun and Node.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ex0b1t** [comment]
Third round triaged. Six findings across both reviewers, all verified against the code first. Five taken as stated, one taken as a documentation fix rather than the suggested code change. Pushed in `f74bf63` and `ced5b68`.

### Taken

**Memory backend keeps settled state on create (cubic, P2).** The real bug of the round. `create` spread its input into storage, so a caller handing over a record that already carried a terminal outcome or resumption metadata kept all of it on a row reporting itself `suspended`. The sqlite backend inserts only the columns it names and never did this, so the two backends stopped being substitutable, which is exactly what the shared contract suite exists to hold. The memory backend now builds the stored record field by field, and a new contract test drives a deliberately polluted input through every backend and asserts it lands clean.

**`NewSuspension` did not actually forbid the fields it omits (cubic, P2).** Correct, and it undercut the type I added last round. A plain `Omit` is satisfied by a full `Suspension` variable, because excess-property checking only fires on a fresh object literal, so the one shape most likely to carry a settled status (a record read back out of the store and re-parked) passed unremarked. Now declared as optional `never` per field. Both test builders return the creation type; nothing else needed changing.

**Non-enumerable named properties on arrays (cubic, P2).** Real, and the same class as the Date fix in the previous round: `Object.keys` does not see them, so they were accepted and dropped on resume. Both checks now read all own names. Regression test added.

**Assert the error code, not just the message (CodeRabbit).** Fair. Three refusal tests matched only message text while their JSDoc promised `RC5042`, so they would have passed on any error mentioning the same identifier. All three now assert the code and keep the message check.

**Teardown ignored `ownsStore` (CodeRabbit).** Latent rather than live, since `MemorySuspensionStore.close()` is a no-op, but the suite's `afterEach` contradicted the ownership contract a test in the same file asserts, and it would bite the moment a caller-supplied backend has a close that does something. Now mirrors the plugin's rule.

### Taken differently

**URL-reserved characters in custom exchange ids (cubic, P1).** The finding is right that my JSDoc overclaimed: it said a suspension id drops into a resume URL without escaping, when only the separator is unreserved. The exchange id it wraps is opaque and an adapter may set `headers["routecraft.id"]` from an upstream message id containing anything.

I did not take the suggested fix of encoding the exchange-id component inside `suspensionIdFor`. That id is the store's primary key and the token's subject; percent-encoding it there means either `#550` double-encodes when it builds the resume link, or every consumer has to know to decode a value that looks like a plain identifier. The id stays opaque, and the encoding obligation now sits explicitly in the JSDoc where the resume-link contract belongs. `~` stays as the separator regardless: it costs nothing and removes one way to get truncated.

`bun run all` green (2342 tests), cross-runtime green under both Bun and Node. The branch is also now merged up to `main` at 0.6.0.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ex0b1t** [comment]
### Recorded as not taken: widening the `better-sqlite3` peer range

CodeRabbit raised widening `better-sqlite3` beyond `^11`. Declined here, recorded so it does not have to be rediscovered.

Majors 12 and 13 exist, and the store's use of the driver is genuinely small: `new Database(path)`, `prepare`, `run`/`get`/`all`, `close`, plus `PRAGMA` statements. Nothing in that surface is plausibly unstable across a major, so the widening is very likely fine on the merits.

The reason it is not in this PR is that a peer range is a compatibility claim, and we would be making one we have not tested. `^11` is what the `adapter-cross-runtime` CI job actually installs and exercises. Declaring `^11 || ^12 || ^13` would tell users we support three majors while proving one, and the failure lands on them at install time, not on us in CI.

**The condition for changing it:** widen the range if and only if the `adapter-cross-runtime` job runs against the widened range first, whether by a matrix over majors or by moving the pin up and testing there. Range and coverage move together, or not at all.

No issue filed. This is a one-line change gated on a CI change, and it belongs to whoever next touches that workflow.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ex0b1t** [comment]
Latest round: three taken, two declined, one escalated as a design decision rather than patched again. Pushed in `fd5c07b`.

### Taken

**sqlite `create()` masks `RC5042` as `RC5044` (P2).** Real, and another backend divergence. `encodePersistable` runs inside the try, so an unpersistable `stepState` surfaced as "failed to persist to the sqlite store" instead of the path-naming boundary error. The memory backend encodes outside any catch and reported it correctly, so the two disagreed on the diagnosis for identical input. `guard()` already has the `isRoutecraftError` passthrough everywhere else in the file, so `create` was the inconsistent one. Verified by probe: both backends now answer `RC5042`.

**`-0` (P3).** Taken despite the low priority, because it completes a rule already implemented twice. `JSON.parse("-0")` yields `-0`, so it reaches a body from an ordinary inbound payload; `JSON.stringify` writes it back as `0` while `structuredClone` preserves it. Same divergence the array-hole and `undefined` normalization exist to prevent.

**"tolerates unhashable adapter state" proved nothing (P3).** Correct. Its values were functions and plain objects, all of which `describable` walks, so no placeholder branch ran and the assertion only showed that hashing returns a digest. Now forces `[opaque]`, `[unhashable]` and `[deep]`.

Also in this commit: the non-enumerable check I added in `fcc5137` compared each own name against a key array, which is quadratic on a large body. Back to linear, with the allocation moved to the throwing path.

### Declined

**Regex literals in `normalizeSource` (P1).** Already documented as a known limitation in the function's JSDoc, with the reason: separating a regex literal from a division operator needs a real tokenizer. The residual risk is narrow, whitespace inside a regex is conventionally written escaped or in a character class, and both survive normalization. Not fixable without the tokenizer, so it stays a stated limitation rather than a patch.

**Explicit `undefined` body (P2).** The shapes differ, the values do not. sqlite drops the key through `JSON.stringify` and reading it back yields `undefined`; memory keeps the key holding `undefined`. Either way `serialized.body === undefined`, so no route or store operation can observe the difference. Not worth a reserved envelope.

### Escalated rather than patched

**Quote-state desync in `normalizeSource` (P3, cubic's own confidence 4).** The finding is right in principle: an apostrophe in a comment flips the quote toggle, and a desynced state could let whitespace inside a later string literal be collapsed. That is the dangerous direction, since it means under-invalidation.

I am not patching it, because this is the third round of findings against the same 50-line hand-rolled scanner (ASI, then regex, now comments), and each patch has been a narrower approximation of a JS tokenizer. Comment tracking would close this one and leave template-literal interpolation open.

The design question is whether `normalizeSource` should exist at all. It buys tolerance of line-ending changes between checkouts. It costs an open-ended class of under-invalidation holes, and under-invalidation is the failure direction the hash exists to prevent, while over-invalidation only re-asks an approval. Hashing raw `Function.prototype.toString()` output deletes the function, closes the class permanently, and moves every remaining failure to the safe direction. The cost is that a CRLF checkout would invalidate parked exchanges, which matters only for a team on mixed line endings, and the build-sensitivity caveat already documented for users covers adjacent ground.

Flagged to Jaco for a decision rather than decided unilaterally. It is a behaviour change to a documented property, not a bug fix.

`bun run all` green (2343 tests), cross-runtime green under both Bun and Node.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/routecraft/src/suspension/hash.ts [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        pending = pending === "\n" || LINE_TERMINATOR.test(char) ? "\n" : " ";
+      }
+      continue;
+    }
+
+    if (pending) {
+      out += pending;
+      pending = "";
+    }
+    if (char === '"' || char === "'" || char === "`") quote = char;
```
[the remark this answers is not in the document]
Comment (ex0b1t) [reply packages/routecraft/src/suspension/hash.ts]: Fixed by deletion in `8779d4f`. `normalizeSource` no longer exists; step source is hashed verbatim.

The finding is correct, and it was the third distinct way to desynchronize that scanner after automatic semicolon insertion and regex literals. Each was fixable only by moving closer to a real JavaScript tokenizer, and each failed in the same direction: a missed change means a parked approval resumes into behaviour its approver never authorized. Patching this one would have closed the comment case and left template-literal interpolation open.

So the class was closed rather than the instance patched. Normalization can only ever fold two distinct sources onto one digest, and every fold is a chance to miss a real change; the opposite error costs an error-channel re-ask, which this design already provides and expects. That asymmetry is now written on `continuationHash` as an explicit design rule so the scanner is not reintroduced by a future contributor trying to make a formatting pass survive.

Landed inside this PR rather than deferred because nothing can park an exchange until #550, so the behaviour change is free today and would cost real approvals at any later point.

The consequence is that the hash now moves on formatting passes and line-ending differences as well as build-emit changes. The configuration reference caveat is rewritten to cover that whole surface in one paragraph, and now recommends pinning line endings with `.gitattributes` alongside the existing build-settings advice. Your sibling P1 on regex literals is closed by the same commit.

---
_Generated by [Claude Code](https://claude.ai/code)_
